inc/Engine/CriticalPath/CriticalCSSSubscriber.php

```diff
@@ -687,39 +594,82 @@ public function async_css( $buffer ) {
+	 * @return string             Modified HTML code with defer CSS.
+	 */
+	private function async_css_tag_update( $tags_match, $buffer ) {
+		$excluded_css = array_flip( $this->critical_css->get_exclude_async_css() );
+		$noscripts    = '';
 		foreach ( $tags_match[0] as $i => $tag ) {
 			// Strip query args.
-			$path = rocket_extract_url_component( $tags_match[2][ $i ], PHP_URL_PATH );
+			$path = rocket_extract_url_component( $tags_match['url'][ $i ], PHP_URL_PATH );
 
 			// Check if this file should be deferred.
 			if ( isset( $excluded_css[ $path ] ) ) {
 				continue;
 			}
 
-			$preload = str_replace( 'stylesheet', 'preload', $tags_match[1][ $i ] );
-			$onload  = preg_replace( '~' . preg_quote( $tags_match[3][ $i ], '~' ) . '~iU', ' as="style" onload=""' . $tags_match[3][ $i ] . '>', $tags_match[3][ $i ] );
-			$tag     = str_replace( $tags_match[3][ $i ] . '>', $onload, $tag );
-			$tag     = str_replace( $tags_match[1][ $i ], $preload, $tag );
-			$tag     = str_replace( 'onload=""', 'onload="this.onload=null;this.rel=\'stylesheet\'"', $tag );
-			$buffer  = str_replace( $tags_match[0][ $i ], $tag, $buffer );
+			$original_media = 'all';
+			$tag            = str_replace( $tags_match['type'][ $i ], ' as="style" ' . $tags_match['type'][ $i ] . ( empty( $tags_match['media'][ $i ] ) ? ' media="print"' : '' ) . ( empty( $tags_match['onload'][ $i ] ) ? ' onload=""' : '' ), $tag );
+			if ( ! empty( $tags_match['media'][ $i ] ) ) {
+				preg_match( '/media\s*=\s*[\'"](?<media>.*)[\'"]/ix', $tags_match['media'][ $i ], $media_match );
+				$original_media = $media_match['media'];
+				$tag            = str_replace( $tags_match['media'][ $i ], 'media="print"', $tag );
+			}
+
+			if ( ! empty( $tags_match['onload'][ $i ] ) ) {
+				$onload_delimiter = substr( $tags_match['onload'][ $i ], 7, 1 );
+				$onload_end       = strpos( $tags_match['onload'][ $i ], $onload_delimiter . ' ' );
+				if ( ! $onload_end ) {
+					$onload_end = strpos( $tags_match['onload'][ $i ], $onload_delimiter . '>' );
+				}
 
+				$onload = substr( $tags_match['onload'][ $i ], 0, $onload_end + 1 );
+				$tag    = str_replace( $onload, 'onload=""', $tag );
+			}
+			$tag        = str_replace( 'onload=""', 'onload="this.media=\'' . $original_media . '\'"', $tag );
+			$buffer     = str_replace( $tags_match[0][ $i ], $tag, $buffer );
 			$noscripts .= '<noscript>' . $tags_match[0][ $i ] . '</noscript>';
 		}
 
 		return str_replace( '</body>', $noscripts . '</body>', $buffer );
 	}
 
+	/**
+	 * Conditions which will allow to defer loading of CSS files.
+	 *
+	 * @return bool True / False if async_css should be applied.
+	 */
+	private function maybe_async_css() {
+		if ( rocket_get_constant( 'DONOTROCKETOPTIMIZE' ) || rocket_get_constant( 'DONOTASYNCCSS' ) ) {
+			return false;
+		}
+		if ( ! $this->options->get( 'async_css' ) ) {
+			return false;
+		}
+		if ( is_rocket_post_excluded_option( 'async_css' ) ) {
+			return false;
+		}
+		return true;
```

Comment:
return the conditional instead of explicitly `false` or `true`.

```suggestion
		return ! is_rocket_post_excluded_option( 'async_css' );
```